### ISSUE TITLE
Eclipse shows "Invalid Overload" and red underline for legal code.

### ISSUE THREAD

**Issue by Sporking** [eclipse-cdt/cdt#189]
For the following code:

```c
#include <cctype>
#include <algorithm>
#include <string>

bool isUpper(const std::string& s)
{
  return std::find_if(s.begin(), s.end(), std::islower) == s.end(); 
}

```

Eclipse CDT (Version: 10.7.0.202206081808, Build id: 20220608-1808) shows a red underline for "std::islower". If I hover over it, Eclipse complains "invalid overload of std::islower".

However, the code compiles and runs fine.

If I remove the 'std' prefix and use "::islower" instead, Eclipse stops complaining.

If I add an actual call to std::islower (e.g. "bool b = std::islower('a');" Eclipse seems to find nothing wrong with that line. It is only when the address of the function is taken, implicitly or explicitly, that Eclipse seems to think there is a problem. For instance, "auto x = std::islower;" shows a red underline, but "auto y = ::islower;" does not.

**Comment by jonahgraham** [comment]
Thanks for the bug report. 

This code does not compile for me in g++, I get:

```
$ g++ -c a.cc
a.cc: In function ‘bool isUpper(const string&)’:
a.cc:7:55: error: no matching function for call to ‘find_if(std::__cxx11::basic_string<char>::const_iterator, std::__cxx11::basic_string<char>::const_iterator, <unresolved overloaded function type>)’
    7 |   return std::find_if(s.begin(), s.end(), std::islower) == s.end();
      |                                                       ^
In file included from /usr/include/c++/10/algorithm:62,
                 from a.cc:2:
/usr/include/c++/10/bits/stl_algo.h:3925:5: note: candidate: ‘template<class _IIter, class _Predicate> _IIter std::find_if(_IIter, _IIter, _Predicate)’
 3925 |     find_if(_InputIterator __first, _InputIterator __last,
      |     ^~~~~~~
/usr/include/c++/10/bits/stl_algo.h:3925:5: note:   template argument deduction/substitution failed:
a.cc:7:55: note:   couldn’t deduce template parameter ‘_Predicate’
    7 |   return std::find_if(s.begin(), s.end(), std::islower) == s.end();
      |                                                       ^

```

Which variant of C++ are you using? CDT is missing some support for C++17 and newer standards.

**Comment by mbooth101** [comment]
The code does not compile even if you pass `-std=c++17` or `-std=c++20` (I'm using GCC 12) -- if you ask CDT to open the declaration, it seems to be trying to resolve `bool std::islower(char, locale)` instead of `int std::islower(char)` 

Maybe this is what the "unresolved overloaded function type" error from GCC is trying to tell me, but I'm not really a C++ guy and I find C++ errors a bit impenetrable :-)

However passing a lambda instead seems to make it compile just fine:
```
return std::find_if(s.begin(), s.end(), [](unsigned char c) {return std::islower(c);}) == s.end(); 
```
And indeed CDT is also happy with it as a lambda, so CDT seems consistent with GCC here.

Maybe @Sporking can tell us more about their compiler version and the compiler flags they are using....

**Comment by Sporking** [comment]
After doing more research, I think that this was the problem:

https://stackoverflow.com/questions/7131858/stdtransform-and-toupper-no-matching-function

Specifically, there is more than one 'std::islower' in the C++ standard library, and when its address is taken without function arguments to disambiguate, the compiler can't decide which one is meant.

I think that the reason that my code was able to successfully compile on Visual Studio 2017 is because they were probably including a different set of C++ standard header files in their includes than the header files I was using with Eclipse (which were from the Cygwin gcc 11 release). So when I compiled, it probably only included one definition of std::islower instead of two that were included when I used Eclipse with gcc header files.

Thanks for  your investigations. I am closing this bug, since at the moment I think that Eclipse was probably behaving correctly.
